### PR TITLE
Check `cedar-policy` SemVer compatibility on pull requests

### PR DESCRIPTION
**Pull request by spinda**
## Description of changes

This adds a GItHub Actions workflow to check changes to `cedar-policy`'s public API against the base branch for SemVer compatibility, using [cargo semver-checks](https://github.com/obi1kenobi/cargo-semver-checks).

Currently this depends on my [fork](https://github.com/spinda/cargo-semver-checks-action/tree/baseline-root-option) of `cargo-semver-checks-action`, with changes I'll send [upstream](https://github.com/obi1kenobi/cargo-semver-checks-action) if this is deemed worth merging.

## Issue #, if available

N/A

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [ ] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).
- [ ] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).
- [ ] A bug fix or other functionality change requiring a patch to `cedar-policy`.
- [x] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)
- [ ] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [ ] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).
- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)
- [ ] Requires updates, but I do not plan to make them in the near future. (Make sure that your changes are hidden behind a feature flag to mark them as experimental.)
- [ ] I'm not sure how my change impacts `cedar-spec`. (Post your PR anyways, and we'll discuss in the comments.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (aaronjeline) [inline .github/workflows/ci.yml:60]: Don't think we can depend on this, can we inline it?

**Review by aaronjeline — CHANGES_REQUESTED** [review]

**Comment by john-h-kastner-aws** [comment]
Is this blocking for a merge into `main`? I don't think that's what we want, at least with our current development practices.

**Comment by spinda** [comment]
I think you can change which status checks are required to pass for merging in the repository settings?

I figure it's better to surface this information at the individual PR level rather than in bulk on preparing to cut a release. At code review time, the results are more actionable: unintended API breakage can be immediately corrected, and intended breakage can be allowed through. Then each merge ends up annotated with whether it breaks SemVer, which could be used for deciding what's safe to cherry-pick into a release. If the checks are put off until it's time to actually cut a release, then fixing detected breakage would involve tracking down the origin of each incompatibility and looping in the responsible parties. Scheduling and other pressures at release-time could make it tempting to take shortcuts or allow things through that we otherwise wouldn't. This was my takeaway from the [Google paper](https://research.google/pubs/pub46576/) on successfully implementing static analysis tools in developer workflows.

**Comment by john-h-kastner-aws** [comment]
Checking semver for each PR is good, but I would not want to be a position where we are regularly overriding an otherwise blocking CI failure due to an expected semver break.

File: .github/workflows/ci.yml
Comment (spinda) [reply .github/workflows/ci.yml:60]: Is the problem that it's pointing to my personal fork right now? If it is, I'm hoping to get that [upstreamed](https://github.com/obi1kenobi/cargo-semver-checks-action) before merging this PR. The author of the upstream action is also the author of `cargo semver-checks`.

If the problem is depending on third-party actions in general: the GitHub Actions documentation is a little spotty, but I believe you can [specify a local path within the repository](https://docs.github.com/en/actions/creating-actions/about-custom-actions#choosing-a-location-for-your-action) for an action, so we ought to be able to inline this.

Comment (aaronjeline) [reply .github/workflows/ci.yml:60]: Third party actions in general yeah

**Comment by spinda** [comment]
Looks like we could set [`continue-on-error`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error) on the `cargo_semver_checks` job to accomplish that.

**Comment by spinda** [comment]
I set `continue-on-error` as above and swapped out the custom action for a few `run:` lines.

**Comment by spinda** [comment]
I tacked on sample commit that breaks SemVer compatibility. How does that look?

**Review by andrewmwells-amazon — APPROVED** [review]

**Comment by andrewmwells-amazon** [comment]
> I tacked on sample commit that breaks SemVer compatibility. How does that look?

LGTM

**Comment by andrewmwells-amazon** [comment]
> Checking semver for each PR is good, but I would not want to be a position where we are regularly overriding an otherwise blocking CI failure due to an expected semver break.
> 
> If there's an option to treat this one check as a warning, then that's good by me.

As I understand it, this will check for symver breaks just for this PR's change. So I don't think we'd have to override the merge continually (as we would if it checked for symver against e.g. the latest tagged commit starting with "v").

Personally I'm fine with this. It's similar to our process for overriding changes that will break cedar-spec. I view this as a sanity check where the failing action reminds you to check that the changelog is modified in this PR before overriding.

**Comment by aaronjeline** [comment]
I think it's at least worth trying. It's not hard to undo if we end up not liking the process.

**Review by aaronjeline — APPROVED** [review]

**Comment by spinda** [comment]
I've made the check optional for normal PRs and mandatory for PRs into `release/*` branches. For normal PRs, the SemVer check is relative to the code in the target branch, so existing SemVer breakage is ignored and only new breakage introduced by the PR should be surfaced. For release PRs, the SemVer check is relative to the latest version published to crates.io <= the version in the PR's `cedar-policy/Cargo.toml`.

This PR should currently be showing up with a failed SemVer check that *doesn't* block merging. I submitted https://github.com/cedar-policy/cedar/pull/172 against `release/2.3.x`, which should be showing up with a failed SemVer check that *does* block merging.

@andrewmwells-amazon @john-h-kastner-aws If this looks good to y'all (I can't confirm from my end whether the checks are properly requiring/not-requiring overrides), then I'll remove the breakage-inducing commit and this'll be ready to merge.

**Review by john-h-kastner-aws — APPROVED** [review]
Seems alright to me.

Two comments the CI script.

**Comment by john-h-kastner-aws** [comment]
After approving I see
image

which looks good to me. It clearly signals that something is failing, but the merge is possible without override.

File: .github/workflows/ci.yml
Comment (andrewmwells-amazon) [reply .github/workflows/ci.yml:85]: I'd either put this in it's own `.sh` file or move the `cargo_semver_checks` to it's own `.yml` file to keep the CI file cleaner. You can see an example in https://github.com/dafny-lang/dafny/blob/master/.github/workflows/publish-release.yml

**Comment by andrewmwells-amazon** [comment]
>If this looks good to y'all (I can't confirm from my end whether the checks are properly requiring/not-requiring overrides), then I'll remove the breakage-inducing commit and this'll be ready to merge.

Seems good to me (modulo comments)

File: .github/workflows/ci.yml
Comment (spinda) [reply .github/workflows/ci.yml]: I moved it into its own `.sh` file.

**Comment by spinda** [comment]
This should be good to go!

**Review by john-h-kastner-aws — APPROVED** [review]

**Comment by spinda** [comment]
Resolved a merge conflict.

**Review by john-h-kastner-aws — APPROVED** [review]

**Pull request merged**
